velox/dwio/dwrf/common/FileMetadata.h

```diff
@@ -0,0 +1,442 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <string>
+
+#include "velox/dwio/common/Common.h"
+#include "velox/dwio/dwrf/common/Common.h"
+#include "velox/dwio/dwrf/common/wrap/dwrf-proto-wrapper.h"
+#include "velox/dwio/dwrf/common/wrap/orc-proto-wrapper.h"
+
+namespace facebook::velox::dwrf {
+
+enum DwrfFormat : uint8_t { kDwrf, kOrc };
+
+class ProtoWrapperBase {
+ protected:
+  DwrfFormat format_;
+  void* impl_;
+
+ public:
+  ProtoWrapperBase() = delete;
+
+  ProtoWrapperBase(DwrfFormat format, void* impl)
+      : format_{format}, impl_{impl} {}
+
+  DwrfFormat format() const {
+    return format_;
+  }
+
+  inline void* rawProtoPtr() const {
+    return impl_;
+  }
+};
+
+class PostScript {
+ public:
+  PostScript(
+      uint64_t footerLength,
+      dwio::common::CompressionKind compression,
+      uint64_t compressionBlockSize,
+      uint32_t writerVersion)
+      : footerLength_{footerLength},
+        compression_{compression},
+        compressionBlockSize_{compressionBlockSize},
+        writerVersion_{static_cast<WriterVersion>(writerVersion)} {}
+
+  explicit PostScript(const proto::PostScript& ps);
+
+  explicit PostScript(const proto::orc::PostScript& ps);
+
+  dwio::common::FileFormat fileFormat() const {
+    return fileFormat_;
+  }
+
+  // General methods
+  uint64_t footerLength() const {
+    return footerLength_;
+  }
+
+  dwio::common::CompressionKind compression() const {
+    return compression_;
+  }
+
+  uint64_t compressionBlockSize() const {
+    return compressionBlockSize_;
+  }
+
+  uint32_t writerVersion() const {
+    return writerVersion_;
+  }
+
+  // DWRF-specific methods
+  StripeCacheMode cacheMode() const {
+    return cacheMode_;
+  }
+
+  uint32_t cacheSize() const {
+    return cacheSize_;
+  }
+
+ private:
+  // General attributes
+  dwio::common::FileFormat fileFormat_ = dwio::common::FileFormat::DWRF;
+  uint64_t footerLength_;
+  dwio::common::CompressionKind compression_ =
+      dwio::common::CompressionKind::CompressionKind_NONE;
+  uint64_t compressionBlockSize_ = dwio::common::DEFAULT_COMPRESSION_BLOCK_SIZE;
+  WriterVersion writerVersion_ = WriterVersion::ORIGINAL;
+
+  // DWRF-specific attributes
+  StripeCacheMode cacheMode_;
+  uint32_t cacheSize_ = 0;
+
+  // ORC-specific attributes
+  uint64_t metadataLength_;
+  uint64_t stripeStatisticsLength_;
+};
+
+class ProtoStripeInformation : public ProtoWrapperBase {
```

Comment:
Can we add "wrapper" to the class name so that people will be more aware of the lifecycle restriction here?

Reply:
Renamed Proto{FieldName} -> {FieldName}Wrapper